src/plugins/plugins_ex.h

```diff
@@ -386,142 +333,201 @@ struct breakpoint_by_pid_searcher
+            delete trap;
             return nullptr;
         }
 
-        attach_plugin_params(trap.get());
-        params.release();
-        return trap.release();
+        traps.push_back(std::move(trap));
+        return traps.back();
     }
 
-    virtual void destroy_trap(drakvuf_t drakvuf, drakvuf_trap_t* trap);
-    drakvuf_trap_t* detach_plugin_params(drakvuf_trap_t* data);
-    static void destroy_plugin_params(drakvuf_trap_t* data);
-
-protected:
-    void attach_plugin_params(drakvuf_trap_t* data);
-
-    template<typename Plugin, typename Params, typename AP = allocator<Params>, typename AT = allocator<drakvuf_trap_t>>
-    bool init_memory(Plugin* plugin,
-                     std::unique_ptr<drakvuf_trap_t>& trap,
-                     std::unique_ptr<Params>& params,
-                     hook_cb_t hook_cb,
-                     const char* trap_name = nullptr,
-                     AP ap = allocator<Params>(),
-                     AT at = allocator<drakvuf_trap_t>())
+    void destroy_plugin_params(drakvuf_trap_t* target)
     {
-        trap.reset(at.allocate(1));
-        if (!trap)
+        std::for_each(traps.begin(), traps.end(), [=](const auto& trap)
         {
-            PRINT_DEBUG("%s. Failed to allocate a memory for trap of %s\n", ERROR_MSG_ADDING_TRAP, trap_name);
-            return false;
-        }
+            if (trap != target)
+                return;
+
+            auto data = static_cast<plugin_data*>(trap->data);
+            if (data->params)
+            {
+                delete data->params;
+                data->params = nullptr;
+            }
+            else
+                PRINT_DEBUG("[PLUGINS_EX] BUG?? params already deleted before");
+        });
+    }
 
-        params.reset(ap.allocate(1, plugin));
-        if (!params)
+    void destroy_trap(drakvuf_trap_t* target)
+    {
+        auto it = std::find(traps.begin(), traps.end(), target);
+        if (it == traps.end())
         {
-            PRINT_DEBUG("%s. Failed to allocate a memory for trap params of %s\n", ERROR_MSG_ADDING_TRAP, trap_name);
-            trap.reset();
-            return false;
+            PRINT_DEBUG("[PLUGINEX] BUG: attempted to destroy non-existant trap");
+            throw -1;
         }
-
-        trap->cb = hook_cb;
-        trap->data = params.get();
-        trap->name = trap_name;
-        trap->type = BREAKPOINT;
-        return true;
+        traps.erase(it);
+        delete_trap(target);
     }
 
-public:
     const output_format_t m_output_format;
 
 private:
-    GSList* m_params;
+    std::vector<drakvuf_trap_t*> traps;
+    drakvuf_t drakvuf;
+
+    void delete_trap(drakvuf_trap_t* target)
+    {
+        // remove trap
+        drakvuf_remove_trap(drakvuf, target, [](drakvuf_trap_t* trap)
+        {
+            delete trap;
+        });
+
+        // delete params
+        destroy_plugin_params(target);
```

Comment:
lambda is executed when it's safe to remove the trap, but be aware that in certain contexts (in destructor?) this is practically guaranteed to be called immediately because `drakvuf->in_callback == 0`. Thus, the trap is destroyed immediately and then we destroy plugin params (inner object) which is undefined behavior

fixup:
```suggestion
        drakvuf_remove_trap(drakvuf, target, [](drakvuf_trap_t* trap)
        {
            destroy_plugin_params(trap);
            delete trap;
        });
```